core: Defer on transitive dependencies for data resources with conditions (#31028)
hashicorp/terraform: website/docs/language/data-sources/index.mdx

## Data Resource Behavior

> Terraform aims to read data resources during the planning phase whenever
> possible, but in order to achieve a correct order of operations Terraform
> defers reading a data resource until the apply phase in the following
> situations:

Reviewer: I'm trying to make this flow better and include the tidbit about announcing it in the plan up here, as I think it's relevant. Ignore this is you don't like this edit....

```suggestion
Terraform reads data resources during the planning phase when possible and announces in the plan when it must defer reading resources until the apply phase to preserve the order of operations. Terraform defers reading data resources in the following situations:
```

Author: I took all of your suggestions here exactly as you wrote them, except that I kept the "but" instead of "and" in the opening sentence here because to me it more clearly read that we're about to describe exceptions to the rule. Please let me know if that doesn't feel right to you!

Author: I'm going to optimistically go ahead and merge this now just because I'm blocking the start of a release process, but I'll be happy to revise this in a subsequent PR if you don't like it! Let me know. :grinning:

Reviewer: ahh sorry that I didn't catch it fast enough. I trust you on this :-) Thank you for letting me know, and I'm glad we were able to get this out!

> * If at least one of the given arguments is a managed resource attribute or

Reviewer: ```suggestion
* At least one of the given arguments is a managed resource attribute or
```

  other value that Terraform cannot predict until the apply step.

> * If the data resource depends directly on a managed resource that itself has

Reviewer: ```suggestion
* The data resource depends directly on a managed resource that itself has
```

  planned changes in the current plan.

> * If the data resource has

Reviewer: ```suggestion
* The data resource has
```

  [custom conditions](#custom-condition-checks)

>   and it depends _directly or indirectly_ on a managed resource that itself

Reviewer: ```suggestion
  and it depends directly or indirectly on a managed resource that 
```

  has planned changes in the current plan.

> See [Data Resource Dependencies](#data-resource-dependencies) below for more
> details on what it means for a data resource to depend on other objects.
> 
> If Terraform needs to defer reading a data resource until the apply phase, it
> will announce that as part of the description of planned changes. Any result
> attribute of such a data resource will be unknown during planning and so cannot
> be used in situations where values must be fully known.

Reviewer: ```suggestion
Refer to [Data Resource Dependencies](#data-resource-dependencies) for 
details on what it means for a data resource to depend on other objects. Any resulting attribute of such a data resource will be unknown during planning, so it cannot be used in situations where values must be fully known. 
```
